common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +177,463 @@
+                            }
+                            line = line.trim();
+                            if (line.length() > 0) {
+                                try {
+                                    Class<?> clazz = Class.forName(line, true, classLoader);
+                                    ExtensionURL extensionUrl = getExtensionURL(clazz);
+                                    extensions.put(extensionUrl, clazz);
+                                } catch (LinkageError | ClassNotFoundException e) {
+                                    LOGGER.warn("load [{}] class fail. {}", line, e.getMessage());
+                                }
                             }
                         }
+                    } catch (Throwable e) {
+                        LOGGER.warn(e.getMessage());
+                    } finally {
+                        IOUtil.close(reader);
                     }
-                } catch (Throwable e) {
-                    LOGGER.warn(e.getMessage());
-                } finally {
-                    IOUtil.close(reader);
                 }
             }
         }
-    }
 
-    /**
-     * init instance
-     *
-     * @param <S>       the type parameter
-     * @param service   the service
-     * @param implClazz the impl clazz
-     * @param argTypes  the arg types
-     * @param args      the args
-     * @return s s
-     * @throws IllegalAccessException the illegal access exception
-     * @throws InstantiationException the instantiation exception
-     * @throws NoSuchMethodException the no such method exception
-     * @throws InvocationTargetException the invocation target exception
-     */
-    protected static <S> S initInstance(Class<S> service, Class implClazz, Class[] argTypes, Object[] args)
-        throws IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException {
-        S s = null;
-        if (argTypes != null && args != null) {
-            // Constructor with arguments
-            Constructor<S> constructor = implClazz.getDeclaredConstructor(argTypes);
-            s = service.cast(constructor.newInstance(args));
-        } else {
-            // default Constructor
-            s = service.cast(implClazz.newInstance());
+        private ExtensionURL getExtensionURL(Class<?> clazz) {
+            String serviceName = null;
+            String typeName = null;
+            Integer priority = 0;
+            Scope scope = Scope.SINGLETON;
+            LoadLevel loadLevel = clazz.getAnnotation(LoadLevel.class);
+            typeName = clazz.getTypeName();
+            if (loadLevel != null) {
+                serviceName = loadLevel.name();
+                priority = loadLevel.order();
+                scope = loadLevel.scope();
+            }
+            ExtensionURL result = new ExtensionURL(serviceName, typeName, priority, scope);
+            //do cache
+            extensionClassUrlMap.put(clazz, result);
+            if (loadLevel != null) {
```

Comment:
Should the `if` statement be `if (serviceName != null)`, or other more suitable one?

Reply:
Since if loadLevel is not null,the name will be required and not null,we can use either loadLevel or serviceName !=null here,to make it easier to understand, changed to (serviceName != null)